docs/secrets.md

```diff
@@ -0,0 +1,44 @@
+Notification endpoints will require credentials in order to communicate with the various external services.
+
+We do not want users to include those credentials in plain text into Flux scripts. We need to provide a mechanism for users load secrets from a safe storage mechanism and provide those secrets to the endpoints.
+
+**Backing Store:** : There are many backing store possibilities, which need not be discussed here.  Any store that provides secure access for key/value lookup may be used by providing an implementation of the KeyLookup  interface in the flux repository.  
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+**Design**
+
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Register an ExternalDependencies struct  (note: it will be straightforward and advantageous to refactor the `InjectFromDependencies` and similar functions into this pattern as well).   
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+func LookupKey(keyObject values.Object) (string, error)
+```
+The keyObject is a values.Object type because: 
+
+- The secret parameter can be passed to the service without being modified
+- The parameter may be validated by the service, minimizing the dependency of the calling function to have prior knowledge of the key lookup protocol. 
```

Comment:
I am not sure I like the values.Object as part of the external interface, as it means that each implementation needs to understand how to extract the key information from the object. 

What about a helper function that knows how to extract the key information from a values.Object consistently, then neither endpoints or implementations of a secret lookup need to know how it works?

Reply:
yeah, same page.  there will need to be a middle layer in this design.  i'm thinking of "service" in abstract terms of something that implements an API.